### DISCUSSION TITLE
[Help need] Error in Entity linking Traing

### DISCUSSION THREAD

**Discussion by aoleiReiz** [explosion/spaCy#7712] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

  File "C:/Users/aolei.bao/PycharmProjects/entity_linking_demo/src/el_train.py", line 271, in train
    sgd=optimizer,
  File "C:\Users\aolei.bao\Envs\ic_new\lib\site-packages\spacy\language.py", line 529, in update
    proc.update(docs, golds, sgd=get_grads, losses=losses, **kwargs)
  File "pipes.pyx", line 1270, in spacy.pipeline.pipes.EntityLinker.update
RuntimeError: [E188] Could not match the gold entity links to entities in the doc - make sure the gold EL data refers to valid results of the named entity recognizer in the `nlp` pipeline.

I used custom trained NER model for entity linking training, also I wrote script to make sure the entity ner model recoginzed has been annotated and contained in Knowledgebase, still getting this error, what else can i do to fix this training error, thanks

## Your Environment

* Operating System: Win10
* Python Version Used: 3.6.8
* spaCy Version Used: 2.3.5
* Environment Information:

**Comment by svlandeg** [comment]
Hi! As a maintenance note - this is a good question for the discussion forum, so I'll move it there. The original issue will be closed, but will link through to the new, open discussion thread.

**Comment by svlandeg** [comment, marked as answer]
Sorry to hear you've been running into issues with this!

> RuntimeError: [E188] Could not match the gold entity links to entities in the doc - make sure the gold EL data refers to valid results of the named entity recognizer in the `nlp` pipeline.

In spaCy v2, this error message typically means that you're feeding training data into the pipeline that can not be matched to entities in the document. Can you provide more details on your approach, preferably a minimal reproducible code snippet? How are you defining your `nlp` pipeline? How did you create your training data?

**Reply by aoleiReiz** [reply]
@svlandeg Hi, thanks for quick replying. 
sample code and data can be found here, if you need more details, let me know, thanks in advance
https://drive.google.com/drive/folders/1S0buJ7Ogl-_Pq0MY8-cIuUg699fd0UE2?usp=sharing 
here is how nlp defined
`nlp = spacy.load(model_dir)`  
`nlp.add_pipe(nlp.create_pipe('sentencizer')) `
`ruler = EntityRuler(nlp, phrase_matcher_attr="LOWER")`  
`patterns = generate_patterns_from_file()`  
`ruler.add_patterns(patterns)`  
`nlp.add_pipe(ruler)`  

 I met another issue during training, here is error details
`spacy\_ml.py:982: RuntimeWarning: invalid value encountered in true_divide
  cosine = (yh * y).sum(axis=1, keepdims=True) / mul_norms
Traceback (most recent call last):
  File "C:/Users/aolei.bao/PycharmProjects/entity_linking_demo/src/help_el.py", line 89, in <module>
    train()
  File "C:/Users/aolei.bao/PycharmProjects/entity_linking_demo/src/help_el.py", line 73, in train
    sgd=optimizer,
  File "C:\Users\aolei.bao\Envs\ic_new\lib\site-packages\spacy\language.py", line 531, in update
    sgd(W, dW, key=key)
  File "thinc\neural\optimizers.pyx", line 200, in thinc.neural.optimizers.Optimizer.__call__
AssertionError`

**Reply by aoleiReiz** [reply]
I upgrade the code to spacy 3 and follow the tutorial, it works fine
